onmetal-controller-manager/controllers/networking/natgateway_controller.go

```diff
@@ -0,0 +1,318 @@
+			}
+		}
+	}
+
+	var (
+		currentIp, currentPort int32 = 0, 0
+		newDestinations        []networkingv1alpha1.NATGatewayDestination
+	)
+	for _, v := range destinations {
+		log.V(2).Info("Start processing destination", "destination", v.Name)
+		if len(v.IPs) != 0 {
+			log.V(2).Info("keep previous port assignment of destination", "destination", v.Name)
+			newDestinations = append(newDestinations, *v)
+			continue
+		}
+
+		log.V(2).Info("find free port for destination", "destination", v.Name)
+		var candidate *networkingv1alpha1.NATGatewayDestinationIP
+		for {
+			if int(currentIp) >= len(natGateway.Status.IPs) {
+				log.V(1).Info("No ports left: Skipping destination", "name", v.Name, "UID", v.UID)
+				r.Eventf(natGateway, corev1.EventTypeWarning, events.ErrorNoPortsLeft, "no ports left: skipping destination %s %s", v.Name, v.UID)
+				candidate = nil
+				break
+			}
+
+			port := currentPort*portsPerNetworkInterface + MIN_PORT
+			candidate = &networkingv1alpha1.NATGatewayDestinationIP{
+				IP:      natGateway.Status.IPs[currentIp].IP,
+				Port:    port,
+				EndPort: port + portsPerNetworkInterface - 1,
+			}
+
+			if _, allocated := used[hash(candidate)]; !allocated {
+				used[hash(candidate)] = true
+				log.V(2).Info("found port range for destination", "destination", v.Name, "start", candidate.Port, "end", candidate.EndPort, "ip", candidate.IP.String())
+				break
+			}
+
+			currentPort++
+			if int(currentPort) > slotsPerIP {
+				currentPort = 0
+				currentIp++
+			}
+		}
+
+		if candidate != nil {
+			v.IPs = []networkingv1alpha1.NATGatewayDestinationIP{*candidate}
+		}
+		newDestinations = append(newDestinations, *v)
+	}
+
+	return newDestinations, int32(len(used)), nil
+}
+
+func (r *NatGatewayReconciler) reconcile(ctx context.Context, log logr.Logger, natGateway *networkingv1alpha1.NATGateway) (ctrl.Result, error) {
+	log.V(1).Info("Reconcile")
+
+	nicSelector := natGateway.Spec.NetworkInterfaceSelector
+	if nicSelector == nil {
+		log.V(1).Info("Network interface selector is empty")
+		return ctrl.Result{}, nil
+	}
+	log.V(1).Info("Network interface selector is present, managing routing")
+
+	if natGateway.Spec.PortsPerNetworkInterface == nil {
+		log.V(1).Info("PortsPerNetworkInterface is empty")
+		return ctrl.Result{}, nil
+	}
+	log.V(1).Info("PortsPerNetworkInterface is present")
+	portsPerNetworkInterface := *natGateway.Spec.PortsPerNetworkInterface
+
+	log.V(1).Info("Finding destinations")
+	destinations, err := r.findDestinations(ctx, log, natGateway)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("error finding destinations: %w", err)
+	}
+	log.V(1).Info("Successfully found destinations", "Destinations", destinations)
+
+	slotsPerIP := int((MAX_PORT - MIN_PORT) / portsPerNetworkInterface)
+	neededIps := int(math.Ceil(float64(len(destinations)) / float64(slotsPerIP)))
+
+	if len(natGateway.Spec.IPs) < neededIps {
+		log.Info(fmt.Sprintf("Got %d ips but %d are needed", len(natGateway.Spec.IPs), neededIps))
```

Comment:
I would suggest using a structured logging here instead of a nested `fmt.Sprintf`.